### PR TITLE
doc/dev: continue writing Testing in the cloud chapter

### PR DESCRIPTION
**Pull request by smithfarm**
This is a work in progress. I'm opening the pull request early so I can remind myself what the name of my wip branch is.

### CODE REVIEW

**Comment by smithfarm** [comment]
@dachary: This is still pretty rough, but you could have a look?

File: doc/dev/index.rst [hunk 1]
```diff
@@ -542,6 +542,33 @@ is deemed to be an integration test. Integration tests are organized into
...
+to that framework from the perspective of a beginning Ceph developer.
+
+The most important thing to understand about teuthology
+-------------------------------------------------------
+
+Teuthology consumes packages. 
+
+It may take some time to understand the significance of this fact, but it
+is `very` significant. It means that automated tests can be conducted on
+multiple platforms using the same packages (RPM, Debian) that can be
```
Comment (theanalyst) [inline doc/dev/index.rst]: nit, probably DEB instead of debian?

Comment (dachary) [inline doc/dev/index.rst]: +1

File: doc/dev/index.rst [hunk 2]
```diff
@@ -542,6 +542,33 @@ is deemed to be an integration test. Integration tests are organized into
 "suites", which are defined in the `ceph-qa-suite repository`_ and run with
 the `teuthology-suite`_ command.
 
+.. _`teuthology-suite`: http://url/of/teuthology-suite/documentation
```
Comment (dachary) [inline doc/dev/index.rst]: there is no documentation other than --help

File: doc/dev/index.rst [hunk 3]
```diff
@@ -542,6 +542,33 @@ is deemed to be an integration test. Integration tests are organized into
...
+
+The `teuthology-suite`_ command is part of the `teuthology framework`_.
+In the sections that follow we attempt to provide a detailed introduction
+to that framework from the perspective of a beginning Ceph developer.
+
+The most important thing to understand about teuthology
```
Comment (dachary) [inline doc/dev/index.rst]: I would leave this out. Having the **Teuthology consumes packages** first is probably a good indication that it is something important.

File: doc/dev/index.rst [hunk 4]
```diff
@@ -778,13 +802,13 @@ As noted in the previous section, most tests are not defined in a single
 yaml file, but rather as a `combination` of files collected from a
 directory tree within the ``suites/`` subdirectory of the `ceph-qa-suite repository`_. 
 
-The set of all tests generated from a given subdirectory of ``suites/``
-is called an "integration test suite", or a "teuthology suite".
+The set of all tests defined by a given subdirectory of ``suites/`` is
```
Comment (dachary) [inline doc/dev/index.rst]: I better like defined also

File: doc/dev/index.rst [hunk 5]
```diff
@@ -793,7 +817,7 @@ The convolution operator, implemented as an empty file called ``%``, tells
 teuthology to construct a test matrix from yaml facets found in
 subdirectories below the directory containing the operator.
 
-The `ceph-disk suite
+For example, the `ceph-disk suite
 <https://github.com/ceph/ceph-qa-suite/tree/master/suites/ceph-disk/>`_ is
```
Comment (dachary) [inline doc/dev/index.rst]: s/master/jewel/ so that there is no risk of that changing later

File: doc/dev/index.rst [hunk 6]
```diff
@@ -841,6 +865,11 @@ An individual test from the ceph-disk suite can be run by adding the
       --suite ceph-disk/basic \
       --filter 'ceph-disk/basic/{distros/ubuntu_14.04.yaml tasks/ceph-disk.yaml}'
 
+.. note: To run a standalone test, ``--suite`` alone is sufficient. If you
```
Comment (dachary) [inline doc/dev/index.rst]: s/standalone/specific/ s/standalone/given/ ? I my mind standalone is roughly the same as a test defined by a single yaml file, which is possible, although uncommon.

File: doc/dev/index.rst [hunk 7]
```diff
@@ -841,6 +865,11 @@ An individual test from the ceph-disk suite can be run by adding the
...
+   want to run a single combination out of a suite that is defined as a
+   directory tree, ``--suite`` must be combined with ``--filter``. This is
+   because the ``--suite`` option understands POSIX relative paths only.
```
Comment (dachary) [inline doc/dev/index.rst]: Hum, now I understand what you're suggesting. An example would be good to illustrate that and avoid confusion.

File: doc/dev/index.rst [hunk 8]
```diff
@@ -1021,6 +1050,53 @@ drive::
 
     $ ceph-workbench ceph-qa-suite --suite dummy
 
+Your first run of `ceph-workbench ceph-qa-suite`_ on a given tenant will
+take a long time to complete because it downloads a VM image. The images
+are cached in FIXME, so they are only downloaded once.
```
Comment (dachary) [inline doc/dev/index.rst]: s/FIXME/OpenStack/

File: doc/dev/index.rst [hunk 9]
```diff
@@ -1021,6 +1050,53 @@ drive::
...
+
+Although ``dummy`` suite does not run any tests, in all other respects it
+behaves just like a teuthology suite and produces some of the same
+artifacts (see `Artifacts produced by teuthology-suite`_).
+
+You can tell whether the ``dummy`` suite completed successfully by FIXME.
```
Comment (dachary) [inline doc/dev/index.rst]: s/FIXME/looking at the pulpito dashboard/ ?

File: doc/dev/index.rst [hunk 10]
```diff
@@ -1021,6 +1050,53 @@ drive::
...
+
+The last line of output looks something like this::
+
+    ssh -i /home/smithfarm/.ceph-workbench/teuthology-myself.pem ubuntu@149.202.168.201
+
+With this command you can access the teuthology node that was just spun
+up. Since the teuthology node is persistent, it may be useful to save this
+command for later reference - for example, by making it an alias::
+
+    alias teuthology="ssh -i . . ."
```
Comment (dachary) [inline doc/dev/index.rst]: I would not advise that. teuthology nodes are semi-persistent. You may keep them for a few days while working on a PR. Or just for one hour if checking something specific.

File: doc/dev/index.rst [hunk 11]
```diff
@@ -1021,6 +1050,53 @@ drive::
...
+---------------------
+
+The standalone test explained in `Reading a standalone test`_ can be run
+with the following command::
+
+    $ ceph-workbench ceph-qa-suite --suite rados/singleton/all/admin-socket.yaml
+
+This will run the suite shown on the current ``master`` branch of
+``ceph/ceph.git``. You can specify a different branch with the ``--ceph``
+option, and even a different git repo with the ``FIXME`` option. (Run
```
Comment (dachary) [inline doc/dev/index.rst]: s/FIXME/--ceph-git-url/

File: doc/dev/index.rst [hunk 12]
```diff
@@ -1021,6 +1050,53 @@ drive::
...
+``teuthology-suite --help`` on the teuthology node for an up-to-date list
+of available options.)
+
```
Comment (dachary) [inline doc/dev/index.rst]: the above with the suggested fixes look good and can be merged. the following lines are WIP I suppose ;-)

File: doc/dev/index.rst (hunk 10 above)
Comment (smithfarm) [inline doc/dev/index.rst]: @dachary I have found that `teuthology` and `packages-repository` remain indefinitely until I kill them manually. Can we add a subcommand to `ceph-workbench` to nuke these?

Comment (dachary) [inline doc/dev/index.rst]: ceph-workbench ceph-qa-suite --teardown does that

File: doc/dev/index.rst (hunk 1 above)
Comment (smithfarm) [inline doc/dev/index.rst]: fixed

File: doc/dev/index.rst (hunk 2 above)
Comment (smithfarm) [inline doc/dev/index.rst]: ok, fixed

File: doc/dev/index.rst (hunk 3 above)
Comment (smithfarm) [inline doc/dev/index.rst]: @dachary I changed it so "Teuthology consumes packages" is the heading.

File: doc/dev/index.rst (hunk 5 above)
Comment (smithfarm) [inline doc/dev/index.rst]: fixed

File: doc/dev/index.rst (hunk 7 above)
Comment (smithfarm) [inline doc/dev/index.rst]: Fixed

File: doc/dev/index.rst (hunk 9 above)
Comment (smithfarm) [inline doc/dev/index.rst]: Unfortunately the `dummy` suite shows up in Pulpito as 1 dead job, so it doesn't really succeed.

File: doc/dev/index.rst (hunk 10 above)
Comment (smithfarm) [inline doc/dev/index.rst]: Fixed.

File: doc/dev/index.rst (hunk 11 above)
Comment (smithfarm) [inline doc/dev/index.rst]: fixed

**Comment by smithfarm** [comment]
@dachary OK, ready.

File: doc/dev/index.rst (hunk 9 above)
Comment (dachary) [inline doc/dev/index.rst]: It should not show as a dead job, if it does, it's a bug.

**Comment by dachary** [comment]
One last FIXME to FIX and it's good :-)

**Comment by smithfarm** [comment]
@dachary FIXME: http://tracker.ceph.com/issues/15020

**Comment by smithfarm** [comment]
@dachary I re-wrote the entire "Run the dummy suite" section.

**Comment by dachary** [comment]
@smithfarm this new section looks good.

You still have

   You can tell whether the ``dummy`` suite completed successfully by FIXME.

**Comment by smithfarm** [comment]
@dachary Fixed in "sqme"

**Comment by dachary** [comment]
@smithfarm :+1: good to squash :-)

**Comment by smithfarm** [comment]
@dachary Squashed.

**Pull request merged**